<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Alec Barrett">

<title>Subtraction_vignette</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="vignette_files/libs/clipboard/clipboard.min.js"></script>
<script src="vignette_files/libs/quarto-html/quarto.js"></script>
<script src="vignette_files/libs/quarto-html/popper.min.js"></script>
<script src="vignette_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="vignette_files/libs/quarto-html/anchor.min.js"></script>
<link href="vignette_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="vignette_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="vignette_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="vignette_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="vignette_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Subtraction_vignette</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Alec Barrett </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="littlebites-subtraction-vignette" class="level1">
<h1>LittleBites Subtraction Vignette</h1>
<p>load in libraries</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(LittleBites)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(bayestestR)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(purrr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Load in the Vignette dataset, these are subsets of the data used in the CeNGEN group paper currently on Biorxiv. Full datasets are downloadable from cengen.org</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>bulk <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">'Data/Bulk_data_5k.tsv'</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>sc <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">'Data/singleCell_reference_5k.tsv'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Load ground truth expression data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>gt_matrix <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">'Data/ubiquitous_non_neuronal/ubituiqtous_and_nonNeuronal_gt_genes_matrix_042222.tsv'</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>train_test <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">'Data/ubiquitous_non_neuronal/ubituiqtous_and_nonNeuronal_gt_genes_split_042222.rds'</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>gt_train <span class="ot">&lt;-</span> gt_matrix[train_test<span class="sc">$</span>training_genes,]</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>gt_test <span class="ot">&lt;-</span> gt_matrix[train_test<span class="sc">$</span>testing_genes,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="calculate-specificity-scores-from-the-single-cell-reference" class="level2">
<h2 class="anchored" data-anchor-id="calculate-specificity-scores-from-the-single-cell-reference">Calculate specificity scores from the single cell reference</h2>
<p>In general, this approach tries to be quite conservative on which genes to subtract out, and how much to subtract. One way to guides this process is to give a gene-level weight based on some metric of how much subtraction should be done at each step.</p>
<p>Here we presume that genes expressed in only 1 cell type are free-game to subtract out, while genes expressed in increasing numbers of cell types should be subtracted with more caution. Thus we can use a tissue specificity score to generate a 0 to 1 weight for each gene, where a score close to 1 indicates effectively perfect specificity (only in intestines for instance), while a score close to 0 indicates ubiquitous expression, and scores in between indicate that the gene is expressed in some smaller proportion of cell types.</p>
<p>I use the Spm measure here, because it is quite sensitive to genes being expressed in even 2 tissue types, and scores fall in value quite quickly, resulting in a more conservative subtraction with less risk of over-fitting.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>cells <span class="ot">&lt;-</span> <span class="fu">colnames</span>(sc) <span class="sc">|&gt;</span> <span class="fu">unique</span>() <span class="sc">|&gt;</span> <span class="fu">sort</span>()</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="do">## calculate specificty scores for each gene using the single cell reference</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>specificity <span class="ot">&lt;-</span> <span class="fu">apply</span>(sc <span class="sc">|&gt;</span> <span class="fu">log1p</span>(), <span class="dv">1</span>, LittleBites<span class="sc">::</span>Spm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="curate-a-list-of-potential-contaminant-cell-types-that-you-want-to-remove" class="level2">
<h2 class="anchored" data-anchor-id="curate-a-list-of-potential-contaminant-cell-types-that-you-want-to-remove">Curate a list of potential contaminant cell types that you want to remove</h2>
<p>In this case I remove only non-neuronal contaminants as neuronal profiles are generally too similar and it can produce major problems during subtraction,</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>contaminants <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'Excretory'</span>, <span class="st">'Glia'</span>, <span class="st">'Hypodermis'</span>, <span class="st">'Intestine'</span>, <span class="st">'Muscle_mesoderm'</span>, <span class="st">'Reproductive'</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>neurons <span class="ot">&lt;-</span> cells[<span class="sc">!</span>cells <span class="sc">%in%</span> <span class="fu">c</span>(contaminants)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="make-a-sample-x-cell_types-matrix-describing-each-cell-type-to-be-modeled-in-the-subtraction-process" class="level2">
<h2 class="anchored" data-anchor-id="make-a-sample-x-cell_types-matrix-describing-each-cell-type-to-be-modeled-in-the-subtraction-process">make a sample x cell_types matrix describing each cell type to be modeled in the subtraction process</h2>
<p>the first column should be the target cell type, and the remaining columns should represent putative contaminants</p>
<p>This matrix will tell the algorithm 1) which cell type we’re targeting to aid in modeling, and 2) which profiles are designated for removal</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>cell_types_matrix <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">colnames</span>(bulk), <span class="cf">function</span>(sample_){</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  cell <span class="ot">&lt;-</span> <span class="fu">str_split_fixed</span>(sample_, <span class="st">'r'</span>, <span class="dv">2</span>)[,<span class="dv">1</span>]</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">c</span>(cell, contaminants))</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">|&gt;</span> <span class="fu">t</span>()</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>cell_types_matrix</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        [,1]  [,2]        [,3]   [,4]         [,5]        [,6]             
AFDr38  "AFD" "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
AFDr39  "AFD" "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
AIMr193 "AIM" "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
AINr185 "AIN" "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
ASIr154 "ASI" "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
ASIr155 "ASI" "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
AVKr110 "AVK" "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
AVKr112 "AVK" "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
AWBr52  "AWB" "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
BAGr119 "BAG" "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
PHAr206 "PHA" "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
PVMr126 "PVM" "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
PVPr3   "PVP" "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
PVQr236 "PVQ" "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
RICr135 "RIC" "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
RIMr224 "RIM" "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
RISr131 "RIS" "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
SMBr196 "SMB" "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
VCr140  "VC"  "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
VCr142  "VC"  "Excretory" "Glia" "Hypodermis" "Intestine" "Muscle_mesoderm"
        [,7]          
AFDr38  "Reproductive"
AFDr39  "Reproductive"
AIMr193 "Reproductive"
AINr185 "Reproductive"
ASIr154 "Reproductive"
ASIr155 "Reproductive"
AVKr110 "Reproductive"
AVKr112 "Reproductive"
AWBr52  "Reproductive"
BAGr119 "Reproductive"
PHAr206 "Reproductive"
PVMr126 "Reproductive"
PVPr3   "Reproductive"
PVQr236 "Reproductive"
RICr135 "Reproductive"
RIMr224 "Reproductive"
RISr131 "Reproductive"
SMBr196 "Reproductive"
VCr140  "Reproductive"
VCr142  "Reproductive"</code></pre>
</div>
</div>
</section>
<section id="run-subtraction" class="level2">
<h2 class="anchored" data-anchor-id="run-subtraction">Run Subtraction!</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>bulk_subtracted <span class="ot">&lt;-</span> <span class="fu">subtraction</span>(<span class="at">bulk =</span> bulk,</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>                               <span class="at">reference =</span> sc,</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>                               <span class="at">cell_types_matrix =</span> cell_types_matrix, </span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>                               <span class="at">training_matrix =</span> gt_train, </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>                               <span class="at">specificity_weights =</span> specificity,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>                               <span class="at">verbose =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "done"</code></pre>
</div>
</div>
</section>
<section id="plotting-the-effects" class="level2">
<h2 class="anchored" data-anchor-id="plotting-the-effects">Plotting the effects</h2>
<p>Here we’ll calculate AUROC scores for each sample individually using the training and test genes to see how the subtraction performed.</p>
<p>First the training genes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>per_sample_bulk_diags <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">colnames</span>(bulk), <span class="cf">function</span>(sample_){</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  cell_type <span class="ot">&lt;-</span> <span class="fu">str_split_fixed</span>(sample_, <span class="st">'r'</span>, <span class="dv">2</span>)[,<span class="dv">1</span>]</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  prediction <span class="ot">&lt;-</span> bulk[train_test<span class="sc">$</span>training_genes, sample_]</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  case <span class="ot">&lt;-</span> gt_train[,cell_type]</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>  diags <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">threshold =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">2</span><span class="sc">**</span><span class="fu">seq</span>(<span class="sc">-</span><span class="dv">17</span>,<span class="dv">12</span>,<span class="fl">0.05</span>)),</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>                  <span class="at">TPR =</span> <span class="fu">map_dbl</span>(threshold, <span class="sc">~</span><span class="fu">get_tpr</span>(prediction, case, .x)),</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>                  <span class="at">FPR =</span> <span class="fu">map_dbl</span>(threshold, <span class="sc">~</span><span class="fu">get_fpr</span>(prediction, case, .x)),</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>                  <span class="at">counts =</span> <span class="st">"raw"</span>)</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>  auroc <span class="ot">&lt;-</span> bayestestR<span class="sc">::</span><span class="fu">auc</span>(diags<span class="sc">$</span>FPR, diags<span class="sc">$</span>TPR)</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(auroc)</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">|&gt;</span> <span class="fu">unlist</span>()</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(per_sample_bulk_diags) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(bulk)</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>per_sample_sub_diags <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">colnames</span>(bulk_subtracted), <span class="cf">function</span>(sample_){</span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>    cell_type <span class="ot">&lt;-</span> <span class="fu">str_split_fixed</span>(sample_, <span class="st">'r'</span>, <span class="dv">2</span>)[,<span class="dv">1</span>]</span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>    prediction <span class="ot">&lt;-</span> bulk_subtracted[train_test<span class="sc">$</span>training_genes, sample_]</span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>    case <span class="ot">&lt;-</span> gt_train[,cell_type]</span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a>    diags <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">threshold =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">2</span><span class="sc">**</span><span class="fu">seq</span>(<span class="sc">-</span><span class="dv">17</span>,<span class="dv">12</span>,<span class="fl">0.05</span>)),</span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a>                    <span class="at">TPR =</span> <span class="fu">map_dbl</span>(threshold, <span class="sc">~</span><span class="fu">get_tpr</span>(prediction, case, .x)),</span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a>                    <span class="at">FPR =</span> <span class="fu">map_dbl</span>(threshold, <span class="sc">~</span><span class="fu">get_fpr</span>(prediction, case, .x)),</span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a>                    <span class="at">counts =</span> <span class="st">"raw"</span>)</span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a>    auroc <span class="ot">&lt;-</span> bayestestR<span class="sc">::</span><span class="fu">auc</span>(diags<span class="sc">$</span>FPR, diags<span class="sc">$</span>TPR)</span>
<span id="cb10-34"><a href="#cb10-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-35"><a href="#cb10-35" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(auroc)</span>
<span id="cb10-36"><a href="#cb10-36" aria-hidden="true" tabindex="-1"></a>  }) <span class="sc">|&gt;</span> <span class="fu">unlist</span>()</span>
<span id="cb10-37"><a href="#cb10-37" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(per_sample_sub_diags) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(bulk_subtracted)</span>
<span id="cb10-38"><a href="#cb10-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-39"><a href="#cb10-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-40"><a href="#cb10-40" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(per_sample_bulk_diags,</span>
<span id="cb10-41"><a href="#cb10-41" aria-hidden="true" tabindex="-1"></a>     per_sample_sub_diags,</span>
<span id="cb10-42"><a href="#cb10-42" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlim =</span> <span class="fu">c</span>(.<span class="dv">7</span>,<span class="dv">1</span>), <span class="at">ylim =</span> <span class="fu">c</span>(.<span class="dv">7</span>,<span class="dv">1</span>),</span>
<span id="cb10-43"><a href="#cb10-43" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">'unaltered bulk training AUROC'</span>,</span>
<span id="cb10-44"><a href="#cb10-44" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">'subtracted bulk training AUROC'</span>)</span>
<span id="cb10-45"><a href="#cb10-45" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="vignette_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>next the reserved testing genes</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>per_sample_bulk_diags_test <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">colnames</span>(bulk), <span class="cf">function</span>(sample_){</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  cell_type <span class="ot">&lt;-</span> <span class="fu">str_split_fixed</span>(sample_, <span class="st">'r'</span>, <span class="dv">2</span>)[,<span class="dv">1</span>]</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  prediction <span class="ot">&lt;-</span> bulk[train_test<span class="sc">$</span>testing_genes, sample_]</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  case <span class="ot">&lt;-</span> gt_test[,cell_type]</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  diags <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">threshold =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">2</span><span class="sc">**</span><span class="fu">seq</span>(<span class="sc">-</span><span class="dv">17</span>,<span class="dv">12</span>,<span class="fl">0.05</span>)),</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>                  <span class="at">TPR =</span> <span class="fu">map_dbl</span>(threshold, <span class="sc">~</span><span class="fu">get_tpr</span>(prediction, case, .x)),</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>                  <span class="at">FPR =</span> <span class="fu">map_dbl</span>(threshold, <span class="sc">~</span><span class="fu">get_fpr</span>(prediction, case, .x)),</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>                  <span class="at">counts =</span> <span class="st">"raw"</span>)</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>  auroc <span class="ot">&lt;-</span> bayestestR<span class="sc">::</span><span class="fu">auc</span>(diags<span class="sc">$</span>FPR, diags<span class="sc">$</span>TPR)</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(auroc)</span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">|&gt;</span> <span class="fu">unlist</span>()</span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(per_sample_bulk_diags_test) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(bulk)</span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a>per_sample_sub_diags_test <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">colnames</span>(bulk_subtracted), <span class="cf">function</span>(sample_){</span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a>  cell_type <span class="ot">&lt;-</span> <span class="fu">str_split_fixed</span>(sample_, <span class="st">'r'</span>, <span class="dv">2</span>)[,<span class="dv">1</span>]</span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true" tabindex="-1"></a>  prediction <span class="ot">&lt;-</span> bulk_subtracted[train_test<span class="sc">$</span>testing_genes, sample_]</span>
<span id="cb11-26"><a href="#cb11-26" aria-hidden="true" tabindex="-1"></a>  case <span class="ot">&lt;-</span> gt_test[,cell_type]</span>
<span id="cb11-27"><a href="#cb11-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-28"><a href="#cb11-28" aria-hidden="true" tabindex="-1"></a>  diags <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">threshold =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">2</span><span class="sc">**</span><span class="fu">seq</span>(<span class="sc">-</span><span class="dv">17</span>,<span class="dv">12</span>,<span class="fl">0.05</span>)),</span>
<span id="cb11-29"><a href="#cb11-29" aria-hidden="true" tabindex="-1"></a>                  <span class="at">TPR =</span> <span class="fu">map_dbl</span>(threshold, <span class="sc">~</span><span class="fu">get_tpr</span>(prediction, case, .x)),</span>
<span id="cb11-30"><a href="#cb11-30" aria-hidden="true" tabindex="-1"></a>                  <span class="at">FPR =</span> <span class="fu">map_dbl</span>(threshold, <span class="sc">~</span><span class="fu">get_fpr</span>(prediction, case, .x)),</span>
<span id="cb11-31"><a href="#cb11-31" aria-hidden="true" tabindex="-1"></a>                  <span class="at">counts =</span> <span class="st">"raw"</span>)</span>
<span id="cb11-32"><a href="#cb11-32" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-33"><a href="#cb11-33" aria-hidden="true" tabindex="-1"></a>  auroc <span class="ot">&lt;-</span> bayestestR<span class="sc">::</span><span class="fu">auc</span>(diags<span class="sc">$</span>FPR, diags<span class="sc">$</span>TPR)</span>
<span id="cb11-34"><a href="#cb11-34" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-35"><a href="#cb11-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(auroc)</span>
<span id="cb11-36"><a href="#cb11-36" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">|&gt;</span> <span class="fu">unlist</span>()</span>
<span id="cb11-37"><a href="#cb11-37" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(per_sample_sub_diags_test) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(bulk_subtracted)</span>
<span id="cb11-38"><a href="#cb11-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-39"><a href="#cb11-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-40"><a href="#cb11-40" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(per_sample_bulk_diags_test,</span>
<span id="cb11-41"><a href="#cb11-41" aria-hidden="true" tabindex="-1"></a>     per_sample_sub_diags_test, </span>
<span id="cb11-42"><a href="#cb11-42" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlim =</span> <span class="fu">c</span>(.<span class="dv">7</span>,<span class="dv">1</span>), <span class="at">ylim =</span> <span class="fu">c</span>(.<span class="dv">7</span>,<span class="dv">1</span>),</span>
<span id="cb11-43"><a href="#cb11-43" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">'unaltered bulk testing AUROC'</span>,</span>
<span id="cb11-44"><a href="#cb11-44" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">'subtracted bulk testing AUROC'</span>)</span>
<span id="cb11-45"><a href="#cb11-45" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="vignette_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>In both cases we can see marked improvement in the accuracy of calling genes expressed (when using ubiquitous and non-neuronal genes).</p>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>